Made it possible to choose which client to use when sending innertube requests and define new clients

iexavl opened the pull request:
At present because youtube is seemingly making some changes to it's APIS, requests made to certain clients might fail.
That's why, in case a request fails, with this one can choose to opt for a different client if the current one is broken. This can either be a registered client in `Clients` , or a custom made one if it has not been added yet to the registered ones. Any methods that use innertube have been updated to use clients specified in the corresponding request. At present, those include requests for video info, search continuation, and one for playlists. Using one of the registered clients is as simple as:
```
        String videoId = "abc12345";
        YoutubeDownloader downloader = new YoutubeDownloader();
        RequestVideoInfo request = new RequestVideoInfo(videoId).client(ClientType.MWEB);
        Response<VideoInfo> response = downloader.getVideoInfo(request);
        VideoInfo video = response.data();
```

sealedtx commented on src/main/java/com/github/kiulian/downloader/parser/ParserImpl.java:
```diff
@@ -439,13 +431,13 @@ private List<PlaylistVideoDetails> parsePlaylistVideos(JSONObject initialData, i
         }
         JSONObject context = initialData.getJSONObject("responseContext");
-        String clientVersion = extractor.extractClientVersionFromContext(context);
+        //String clientVersion = extractor.extractClientVersionFromContext(context);
```
could be removed

sealedtx commented on src/main/java/com/github/kiulian/downloader/parser/ParserImpl.java:
```diff
@@ -473,28 +465,33 @@ private void populatePlaylist(JSONObject content, List<PlaylistVideoDetails> vid
                         .fluentPut("clickTrackingParams", ctp));
 
+
+//        JSONObject body = new JSONObject()
```
could be removed

sealedtx commented on src/main/java/com/github/kiulian/downloader/parser/ParserImpl.java:
```diff
@@ -727,18 +724,22 @@ private SearchResult parseHtmlSearchResult(String url) throws YoutubeException {
                 .fluentPut("clickTracking", new JSONObject()
                         .fluentPut("clickTrackingParams", continuation.clickTrackingParameters()));
 
+//        JSONObject body = new JSONObject()
```
could be removed

sealedtx commented:
@iexavl what do you think about removing `ClientTraits` and keeping only `ClientType`. As you already mentioned we don't have clear definition what min/max qualities stands for each client and this might be updated by youtube anytime, someone might get confused if this library won't be updated as well. Other than this - looks good to me. Thank you

sealedtx commented:
@iexavl ping me when its ready to be merged

sealedtx commented:
@iexavl is it difficult to run this test again? I'd like to see also list of available itags (quality formats) for each of client

iexavl commented:
> @iexavl is it difficult to run this test again? I'd like to see also list of available itags (quality formats) for each of client

[byAudioQuality.txt](https://github.com/user-attachments/files/17234817/byAudioQuality.txt)
[byCumulativeAV.txt](https://github.com/user-attachments/files/17234818/byCumulativeAV.txt)
[bySuccessAndSpeed.txt](https://github.com/user-attachments/files/17234819/bySuccessAndSpeed.txt)
[byVideoQuality.txt](https://github.com/user-attachments/files/17234820/byVideoQuality.txt)
[fullResults.txt](https://github.com/user-attachments/files/17234821/fullResults.txt)

Seems I was wrong about ANDROID_TV. I ran the tests multiple times and TV_UNPLUGGED_CAST always came out on top. 
The videos I ran it with were cherrypicked 4 videos 2160p60 and one 4320p. All the files are sorted as follows:

fullResults.txt : these are just the full test results. Not sorted in any particular order
byCumulativeAV.txt : these results are sorted by the ordinals of the video qualities added to the ordinals of the audio qualities for each video.
byAudioQuality.txt: sorted by the audio qualities
byVideoQuality.txt: sorted by video qualities
bySuccessAndSpeed: sorted by the success rate of the client first and the speed of the client second.

The cumulativeAV, video and audio qualities are sorted solely by their quality values, reliability and speed aren't included there.
I can throw that into the equation and perhaps throw in a couple of function for getting clients based on cumulative AV, video and audio qualities? It's also important to keep in mind that the data set is still relatively small. I did find this though: https://docs.invidious.io/installation/#docker-compose-method-production
And perhaps if I let the test run for a while until I get flagged as a bot I can run this and send all requests along with the po_token. After that I can maybe just take a playlist with like 50 videos and let the test run.

sealedtx commented:
@iexavl greatly appreciate your work! Can you share results in format?

`client_name, success_rate, avg_speed, total_formats_count, [video_format1, ..., audio_format1 ..., audio+video_format_2]`


also would be good to add definitions for unknown itags

iexavl commented:
> @iexavl greatly appreciate your work! Can you share results in format?
> 
> `client_name, success_rate, avg_speed, total_formats_count, [vid_480p, vid_1080p, ..., audio_medium, audio_low..., audio+vid_1080p]`
> 
> also would be good to add definitions for unknown itags

[requestedFormatting.txt](https://github.com/user-attachments/files/17243876/requestedFormatting.txt)
I added the missing Itags. Well, the ones that were reported to me anyway. The results are formatted exactly as you specified 
(the avg_speed is in milliseconds) and are sorted first by success rate, then by cumulative AV, then by speed.

sealedtx commented:
@iexavl seems like WEB needs to be top priority?

iexavl commented:
@sealedtx is there a reason why the jsUrl needs to be extracted from the html response? I looked at the requests youtube was sending and I saw a constant js address. I changed the parseFormats call in parseVideoAndroid to this:
`formats = parseFormats(playerResponse, "https://www.youtube.com/s/player/96d06116/player_ias.vflset/en_US/base.js", clientVersion);`

and it seems to be working.

iexavl commented:
> > @sealedtx is there a reason why the jsUrl needs to be extracted from the html response? I looked at the requests youtube was sending and I saw a constant js address. I changed the parseFormats call in parseVideoAndroid to this: `formats = parseFormats(playerResponse, "https://www.youtube.com/s/player/96d06116/player_ias.vflset/en_US/base.js", clientVersion);`
> > and it looks like it's working. With that when there's a cipher it doesn't give those errors, and the web clients seem to be working fine.
> 
> I can't remember exactly, I believe it is updated from time to time. You can try make default value for jsUrl if it fails to find one, thats would be ok
> 
> Please, do this change, run final test and get the top 1 priority client, finalize the code and I'll merge it

I have changed the ordering according to the latest test results, which are these:
[byRequestedFormatting.txt](https://github.com/user-attachments/files/17262721/byRequestedFormatting.txt)

I still set the default client to ANDROID_VR. The main reason for that is because it doesn't require a cipher unlike pretty much all of the WEB clients. Also, the youtubei player requests never returns a jsUrl at least in my experience. It has to be obtained from the html page `https://www.youtube.com/watch?v`. So I made it if ciphering is required, a request to an arbitrary video is made (in this case this one: HHbt6rPuC3M) so the jsUrl can be obtained, after that it is cached and the request isn't sent again for requests made via clients. That way if the url changes it won't need a source update.

sealedtx commented on videos/3c421297-0048-43dd-81f0-60fc80c0600d.mp4:
what is this?

sealedtx commented on src/main/java/com/github/kiulian/downloader/parser/ParserImpl.java:
```diff
@@ -36,18 +37,49 @@
 public class ParserImpl implements Parser {
     private static final String ANDROID_APIKEY = "AIzaSyAO_FJ2SlqU8Q4STEHLGCilw_Y9_11qcW8";
     private static final String BASE_API_URL= "https://www.youtube.com/youtubei/v1";
+        static class InvalidJsUrlException extends YoutubeException.CipherException{
```
please, move this to `YoutubeException,java` together with other exceptions

sealedtx commented on src/main/java/com/github/kiulian/downloader/parser/ParserImpl.java:
```diff
@@ -105,7 +137,21 @@ private VideoInfo parseVideoAndroid(String videoId, YoutubeCallback<VideoInfo> c
 
-            } catch (YoutubeException e) {
+            }catch (InvalidJsUrlException e){
+                JSONObject playerConfig = downloadPlayerConfig("HHbt6rPuC3M",callback);
```
hardcoded value

sealedtx commented on src/main/java/com/github/kiulian/downloader/parser/ParserImpl.java:
```diff
@@ -105,7 +137,21 @@ private VideoInfo parseVideoAndroid(String videoId, YoutubeCallback<VideoInfo> c
+                String jsUrl;
+                try {
+                    jsUrl = extractor.extractJsUrlFromConfig(playerConfig, videoId);
+                } catch (YoutubeException ex) {
```
why do we need here another catch? it will just trigger `callback.onError(ex);` twice with same exception

sealedtx replied:
@iexavl what about this? seems like unnecessary call `callback.onError(ex);`

sealedtx commented:
@iexavl left comments to fix

sealedtx commented:
@iexavl I'll merge this as it it, if you don't respond until tomorrow, Will fix error callback myself. Thank you very much for your efforts

iexavl commented:
> @iexavl I'll merge this as it it, if you don't respond until tomorrow, Will fix error callback myself. Thank you very much for your efforts

I suppose I'll look forward to it. I am just not sure what you mean so its probably for the best that you do it. Feel free to merge whenever.
